Patch for the Khronos-distributed http://www.khronos.org/registry/cl/ ICD Loader
to make it usable for regression testing in pocl development:
- Allow overriding the .icd search path with env OCL_ICD_VENDORS (like ocl-icd does)
- Do not fail in case the directory contains non .icd files.
- Allow paths without '/' suffix.

--- icd/icd_linux.c	2012-08-06 09:53:39.000000000 +0300
+++ icd-modifier/icd_linux.c	2013-02-26 14:11:21.494293046 +0200
@@ -57,7 +57,11 @@
 {
     DIR *dir = NULL;
     struct dirent *dirEntry = NULL;
-    char *vendorPath = "/etc/OpenCL/vendors/";
+    char *vendorPath = getenv("OCL_ICD_VENDORS");
+    if (vendorPath == NULL || strncmp(vendorPath, "", 1) == 0)
+    {
+        vendorPath = "/etc/OpenCL/vendors/";
+    }
 
     // open the directory
     dir = opendir(vendorPath);
@@ -85,21 +89,21 @@
                 // make sure the file name ends in .icd
                 if (strlen(extension) > strlen(dirEntry->d_name) )
                 {
-                    break;
+                    continue;
                 }
                 if (strcmp(dirEntry->d_name + strlen(dirEntry->d_name) - strlen(extension), extension) ) 
                 {
-                    break;
+                    continue;
                 }
 
                 // allocate space for the full path of the vendor library name
-                fileName = malloc(strlen(dirEntry->d_name) + strlen(vendorPath) + 1);
+                fileName = malloc(strlen(dirEntry->d_name) + strlen(vendorPath) + 2);
                 if (!fileName) 
                 {
                     KHR_ICD_TRACE("Failed allocate space for ICD file path\n");
                     break;
                 }
-                sprintf(fileName, "%s%s", vendorPath, dirEntry->d_name);
+                sprintf(fileName, "%s/%s", vendorPath, dirEntry->d_name);
 
                 // open the file and read its contents
                 fin = fopen(fileName, "r");
